src/protocols/rdp/guac_rdpdr/rdpdr_printer.c

```diff
@@ -141,18 +143,25 @@ static void guac_rdpdr_device_printer_announce_handler(guac_rdpdr_device* device
     Stream_Write(output_stream, "PRN1\0\0\0\0", 8); /* DOS name */
 
     /* Printer data */
-    Stream_Write_UINT32(output_stream, 24 + GUAC_PRINTER_DRIVER_LENGTH + GUAC_PRINTER_NAME_LENGTH);
+    int settings_length = guac_utf8_strlen(device->device_name);
+    int printer_name_length = (settings_length + 1) * 2;
+    char* printer_name = malloc(printer_name_length);
+    guac_rdp_utf8_to_utf16((const unsigned char*)device->device_name, 
+            settings_length, printer_name, printer_name_length);
+    printer_name[printer_name_length - 2] = '\0';
+    printer_name[printer_name_length - 1] = '\0';
+    Stream_Write_UINT32(output_stream, 24 + GUAC_PRINTER_DRIVER_LENGTH + printer_name_length);
     Stream_Write_UINT32(output_stream,
               RDPDR_PRINTER_ANNOUNCE_FLAG_DEFAULTPRINTER
             | RDPDR_PRINTER_ANNOUNCE_FLAG_NETWORKPRINTER);
     Stream_Write_UINT32(output_stream, 0); /* reserved - must be 0 */
     Stream_Write_UINT32(output_stream, 0); /* PnPName length (PnPName is ultimately ignored) */
     Stream_Write_UINT32(output_stream, GUAC_PRINTER_DRIVER_LENGTH); /* DriverName length */
-    Stream_Write_UINT32(output_stream, GUAC_PRINTER_NAME_LENGTH);   /* PrinterName length */
+    Stream_Write_UINT32(output_stream, printer_name_length);   /* PrinterName length */
     Stream_Write_UINT32(output_stream, 0);                          /* CachedFields length */
 
     Stream_Write(output_stream, GUAC_PRINTER_DRIVER, GUAC_PRINTER_DRIVER_LENGTH);
-    Stream_Write(output_stream, GUAC_PRINTER_NAME,   GUAC_PRINTER_NAME_LENGTH);
+    Stream_Write(output_stream, printer_name, printer_name_length);
```

Comment:
We may need to do more with respect to verifying the necessary space is actually available. Looking into the dynamic allocation of the stream buffer, we're currently allocating a static 256 bytes to contain all RDPDR devices:

https://github.com/apache/guacamole-server/blob/344ed4f42ea715a2e0e7dfd09867964ecfc320c2/src/protocols/rdp/guac_rdpdr/rdpdr_messages.c#L126

This was sufficient when the length of device names was known and this length could not possible be exceeded even if both drive and printer were enabled, but device names are now arbitrary.

Enforcing length limits on device names, truncating the names, etc. would be one possibility. Dynamically calculating the necessary space would be another (though more complex).

Reply:
Hmmm....okay.  While enforcing limits and truncating names might be easier, I'm a little skeptical it actually will be any easier, and think probably dynamically calculating the necessary space is the way to go.  I'll see what I can figure out, but I have a feeling I'll be needing some guidance on the best way to accomplish this.

Reply:
Okay, took an initial stab at a calculation...I'm sure it needs some work.

Reply:
I think some of the complexity we're dealing with is due to the concept of the `announce_handler` and overlapping responsibilities between the devices and the top-level handling of RDPDR and the announce packet.

What if, instead of an `announce_handler`, we:

* Stored the device type (`RDPDR_DTYP_PRINT`, etc.) within the `guac_rdpdr_device`
* Stored the arbitrary device data and the length of that data within the `guac_rdpdr_device`
* Relied upon that within `guac_rdpdr_send_client_device_list_announce_request()`, such that it's truly only the responsibility of that function to assemble the announce packet

I think that should simplify things greatly.

If it helps, I'd be glad to take a stab at the above refactor against the current codebase, though I'm not sure how painful the rebases of these name-related PRs would be after that.

Reply:
Sounds good - I'll be happy to have a go at it, first, and see if I can figure out how to do it - I'll let you know if I need some help.

Reply:
Okay, a couple of clarification questions/concerns, here:
- For tracking the size of arbitrary data within the `guac_rdpdr_device`, are you thinking all of the data - name, type, and then anything in the `data` field?  Or just the size of the `data` field?
- The only potential issue with getting rid of `announce_handler`s entirely is that the complexity of the printer one is significantly larger than that of the filesystem one.  I can still do this, it just means that `guac_rdpdr_send_client_device_list_announce_request()` will end up with either some `if...else if` statements or a `switch` statement to handle initialization that goes with each of the devices.  Or did you envision keeping the `announce_handler`s in some form, just moving a bunch of the code over to the `guac_rdpdr_send_client_device_list_announce_request()` function?

Reply:
Well, I've taken a stab at the implementation of this, as seen in the latest commit.  It feels far, far away from anywhere near good at this point, but you can take a look and let me know if I'm on the right track and make suggestions.  The whole double loop thing just feels wrong, but I'm not sure how else to get the number of bytes before opening the stream and then actually write the stream data.